[openwebnet] Add support for Alarm (WHO=5)

mvalla opened the pull request:
This PR adds support for Alarm (WHO=5) to the OpenWebNet binding.
With this enhancement it will be possible to discover alarm-related devices in the system and read status information from BTicino Alarm system and zones.
Fixes #12569

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -409,4 +409,53 @@
+		<item-type>Switch</item-type>
+		<label>System State</label>
+		<description>Alarm system is active (ON) or inactive (OFF) (read only)</description>
+		<state readOnly="true" pattern="%s"/>
```
No need to define a pattern for a switch channel type.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -409,4 +409,53 @@
+		<item-type>Switch</item-type>
+		<label>Armed State</label>
+		<description>Alarm is armed (ON) or disarmed (OFF) (read only)</description>
+		<state readOnly="true" pattern="%s"/>
```
No need to define a pattern for a switch channel type.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -409,4 +409,53 @@
+		<item-type>Switch</item-type>
+		<label>Network State</label>
+		<description>Alarm network state (ON = network ok, OFF = no network) (read only)</description>
+		<state readOnly="true" pattern="%s"/>
```
No need to define a pattern for a switch channel type.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -409,4 +409,53 @@
+		<item-type>String</item-type>
+		<label>Battery State</label>
+		<description>Alarm battery state (read only)</description>
+		<state>
```
You should add readonly.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -409,4 +409,53 @@
+		</state>
+	</channel-type>
+
+	<channel-type id="zoneAlarm">
```
Please update the file openwebnet.properties in i18n folder.
Hint: run `mvn i18n:generate-default-translations`

mvalla replied:
as it was expected the current event options default translations (e.g. `channel-type.openwebnet.cenButtonEvent.event.option.START_PRESS`) get deleted now from the script as it dos not support them yet.
Shall I delete them?

lolodomo replied:
I would suggest to apply only the required changes for that PR,

mvalla replied:
i would prefer to remove those commented translations, until event options are supported by the script, otherwise they would be removed each time the script is executed forcing to re-add them manually each time

lolodomo replied:
Ok, fine.

lolodomo reviewed (changes requested):
Review part 1 of 2.
Remains to review: the new thing handler and README.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    private void updateAlarmSystemState(WhatAlarm w) {
+        if (w == Alarm.WhatAlarm.SYSTEM_ACTIVE) {
+            updateState(CHANNEL_ALARM_SYSTEM_STATE, OnOffType.ON);
+        } else {
+            updateState(CHANNEL_ALARM_SYSTEM_STATE, OnOffType.OFF);
+        }
```
Just a suggestion: you could do it simply in one line:
```
updateState(CHANNEL_ALARM_SYSTEM_STATE, OnOffType.from(w == Alarm.WhatAlarm.SYSTEM_ACTIVE));
```
Same in your other methods.

lolodomo commented on bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:
```diff
@@ -0,0 +1,249 @@
+
+    @Override
+    protected void refreshDevice(boolean refreshAll) {
+        if (refreshAll) {
```
No need to check that thing type is THING_TYPE_BUS_ALARM_SYSTEM there ?

mvalla replied:
in fact no: if refreshAll is true, the state for all alarm devices (both central unit and zones) will be performed by asking a complete alarm system status

lolodomo commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -157,83 +160,97 @@ Temperature sensors can be configured defining a `bus_thermo_sensor` Thing with
+| `network`                    | `bus_alarm_system`                     | Switch      | Alarm system network state (`ON` = network ok, `OFF` = no network)    |      R      |
+| `battery`                    | `bus_alarm_system`                     | String      | Alarm system battery state (`OK`, `FAULT`, `UNLOADED`)                |      R      |
+| `armed`                      | `bus_alarm_system`, `bus_alarm_zone`   | Switch      | Alarm system or zone is armed (`ON`) or disarmed (`OFF`)              |      R      |
+| `alarm`                      | `bus_alarm_zone`                       | String      | Current alarm for the zone  (`NO_ALARM`, `INTRUSION`, `TAMPERING`, `ANTI_PANIC`) |      R      |
```
It is not "NO_ALARM" but rather "SILENT".

lolodomo commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -309,7 +326,11 @@ Bridge openwebnet:bus_gateway:mybridge "MyHOMEServer1" [ host="192.168.1.35", pa
       bus_dry_contact_ir            LR_IR_sensor         "Living Room IR Sensor"    [ where="399" ]
       
-      bus_aux		                Alarm_activation	 "Alarm activation"         [ where="4"   ]
+      bus_aux		                 Alarm_activation	 "Alarm activation"         [ where="4" ]
+      bus_aux		                 Alarm_deactivation	 "Alarm deactivation"       [ where="5" ]
```
You could remove one TAB after bus_aux to have a good alignment.

lolodomo commented on bundles/org.openhab.binding.openwebnet/README.md:
```diff
@@ -467,6 +502,34 @@ when
+when
+    Item iAlarmSystemArmed received update
+then
+    if(iAlarmSystemArmed.state == ON){
+        iAlarmSystemArmedControl.postUpdate(ON)
+    }
+    else if(iAlarmSystemArmed.state == OFF){
```
More reliable to use `newState` instead of `iAlarmSystemArmed`

lolodomo reviewed (changes requested):
Review part 2 of 2

lolodomo commented:
@mvalla :any chance to finish today before the new milestone generation?

mvalla commented:
Actually after some tests I think is better to rename some channels. I will do that following days and for now add ness more testing tag. So better to target next milestone

mvalla commented:
@lolodomo after some tests with users I changed a channel definition to better reflect the active state of an alarm zone (only the alarm system can be armed/disarmed, not each zone). 
I also corrected few bugs and cleaned up the README.
Can you provide another round of review?

lolodomo reviewed (approved):
LGTM, thank you
